### PR TITLE
Ignore AlreadyClosedException on background merges

### PR DESCRIPTION
**Pull request by arteam**
If we encounter an `AlreadyClosedException` when doing background merges, it means the service is being shut down. We can safely ignore the exception and avoid failing the index writer and creating noise in logs.

We also convert `java.nio.channels.ClosedChannelException` to `AlreadyClosedException` if we try to access a closed cache file during a merge.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/engine/ElasticsearchConcurrentMergeScheduler.java [hunk 1]
```diff
@@ -132,6 +133,8 @@ protected void doMerge(MergeSource mergeSource, MergePolicy.OneMerge merge) thro
         try {
             beforeMerge(onGoingMerge);
             super.doMerge(mergeSource, merge);
+        } catch (AlreadyClosedException ignore) {
+            // Intentially ignored in order not to fail `IndexWriter` when we already shutting down the service.
```
Comment (idegtiarenko) [inline server/src/main/java/org/elasticsearch/index/engine/ElasticsearchConcurrentMergeScheduler.java:137]: I wonder if we should have at least a `logger.debug` here.It proved to be helpful while debugging unexpected segment count

File: x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBytes.java [hunk 2]
```diff
@@ -339,7 +345,12 @@ public int write(ByteBuffer src, int position) throws IOException {
             assert position % PAGE_SIZE == 0;
             assert src.remaining() % PAGE_SIZE == 0;
             checkOffsets(position, src.remaining());
-            int bytesWritten = fileChannel.write(src, pageStart + position);
+            int bytesWritten;
+            try {
+                bytesWritten = fileChannel.write(src, pageStart + position);
+            } catch (ClosedChannelException e) {
+                throw new AlreadyClosedException("Cache has been closed", e);
```
Comment (ywangd) [inline x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBytes.java:352]: Is this still going to be caught by SearchIndexInput which leads to fallback to reading from the object store?

**Review by henningandersen — CHANGES_REQUESTED** [review]
I would like to instead wait for shards to close - including the async-close part - in `IndicesService.doStop()`. From other conversations, I understand the case where this does not happen to be when an index/shard is first closed/removed, then the `IndicesService` is stopped - we'll then not wait for the first shard's closure. This seems like an oversight more than as expected. I'll note that the engine close is async, including waiting for open-refs (i.e., some ops completion). Not waiting for engine (async) close before taking other components down could lead to all sorts of exceptions, not only during merging.

**Comment by arteam** [comment]
@elasticmachine update branch

**Comment by arteam** [comment]
@elasticmachine update branch

**Pull request closed without merging**